File: TimberbornTerrainGenerator/Terrain/CaveGenerator.cs
using TimberbornTerrainGenerator.Config;
using TimberbornTerrainGenerator.Utils;

namespace TimberbornTerrainGenerator.Terrain;

public class CaveGenerator
{
    private FastNoiseLite? _noise;

    public void GenerateCaveSystem(VoxelGrid grid, GeneratorConfig config)
    {
        var settings = config.Caves;
        if (!settings.Generate) return;

        _noise = new FastNoiseLite(config.Seed + 1000);
        _noise.SetNoiseType(FastNoiseLite.NoiseType.Cellular);
        _noise.SetFrequency(0.05f);

        var random = new Random(config.Seed);

        // Generate worm tunnels
        for (int i = 0; i < settings.WormCount; i++)
        {
            GenerateWorm(grid, random, settings, config.MaxHeight);
        }

        // Apply 3D noise for cavern chambers
        ApplyCavernNoise(grid, config);
    }

    private void GenerateWorm(VoxelGrid grid, Random random, CaveSettings settings, int maxHeight)
    {
        // Start point: underground, random X/Z
        var position = new Vector3(
            random.Next(grid.Width),
            random.Next(settings.MinCaveDepth, maxHeight / 2),
            random.Next(grid.Depth)
        );

        // Random initial direction (mostly horizontal)
        var direction = new Vector3(
            random.NextSingle() * 2 - 1,
            (random.NextSingle() - 0.5f) * 0.3f, // Less vertical
            random.NextSingle() * 2 - 1
        ).Normalized();

        for (int step = 0; step < settings.WormLength; step++)
        {
            // Carve sphere at current position
            CarveSphere(grid, position, settings.WormRadius);

            // Smoothly vary direction
            direction = RotateRandomly(direction, random, 15f);
            position += direction * (settings.WormRadius * 0.6f);

            // Keep within bounds
            if (!IsValidWormPosition(grid, position, settings.MinCaveDepth))
                break;
        }
    }

    private void CarveSphere(VoxelGrid grid, Vector3 center, float radius)
    {
        int iRadius = (int)Math.Ceiling(radius);
        var centerInt = new Vector3Int(
            (int)Math.Round(center.X),
            (int)Math.Round(center.Y),
            (int)Math.Round(center.Z)
        );

        for (int dx = -iRadius; dx <= iRadius; dx++)
        for (int dy = -iRadius; dy <= iRadius; dy++)
        for (int dz = -iRadius; dz <= iRadius; dz++)
        {
            var offset = new Vector3Int(dx, dy, dz);
            if (offset.Magnitude <= radius)
            {
                var pos = centerInt + offset;
                if (grid.IsInBounds(pos))
                    grid[pos] = VoxelType.Air;
            }
        }
    }

    private Vector3 RotateRandomly(Vector3 direction, Random random, float maxAngle)
    {
        // Apply random rotation (simplified)
        var noise = new Vector3(
            (random.NextSingle() - 0.5f) * 0.3f,
            (random.NextSingle() - 0.5f) * 0.2f,
            (random.NextSingle() - 0.5f) * 0.3f
        );

        return (direction + noise).Normalized();
    }

    private bool IsValidWormPosition(VoxelGrid grid, Vector3 position, int minDepth)
    {
        return position.X >= 0 && position.X < grid.Width &&
               position.Y >= minDepth && position.Y < grid.Height &&
               position.Z >= 0 && position.Z < grid.Depth;
    }

    private void ApplyCavernNoise(VoxelGrid grid, GeneratorConfig config)
    {
        if (_noise == null) return;

        var settings = config.Caves;

        for (int x = 0; x < grid.Width; x++)
        for (int y = settings.MinCaveDepth; y < config.MaxHeight - 3; y++)
        for (int z = 0; z < grid.Depth; z++)
        {
            var pos = new Vector3Int(x, y, z);
            if (grid[pos] != VoxelType.Solid) continue;

            float noiseValue = _noise.GetNoise(x, y, z);
            if (noiseValue > settings.CaveNoiseThreshold)
            {
                grid[pos] = VoxelType.Air;
            }
        }
    }

    public void CreateCaveEntrances(VoxelGrid grid)
    {
        // Find cave openings near surface
        for (int x = 0; x < grid.Width; x++)
        for (int z = 0; z < grid.Depth; z++)
        {
            // Scan downward from top
            for (int y = grid.Height - 1; y > 0; y--)
            {
                var pos = new Vector3Int(x, y, z);
                var below = new Vector3Int(x, y - 1, z);

                // Found transition from solid to cave
                if (grid[pos] == VoxelType.Solid && grid[below] == VoxelType.Air)
                {
                    // Remove ceiling blocks to create entrance
                    grid[pos] = VoxelType.Air;
                    break;
                }
            }
        }
    }
}

File: TimberbornGenerator/src/terrain/CaveGenerator.ts
import FastNoiseLite from 'fastnoise-lite';
import { VoxelGrid } from '../core/VoxelGrid';
import { GeneratorConfig, Vector3Int, VoxelType } from '../types';

/**
 * CaveGenerator - Creates cave systems using worm tunnels and 3D noise
 *
 * 🔥 WASM CANDIDATE: Cave carving involves intensive 3D iteration and noise sampling.
 * A Rust WASM implementation could provide:
 * - 5-10x faster cave generation
 * - Parallel tunnel carving
 * - SIMD for bulk voxel clearing
 *
 * TypeScript is acceptable for small cave systems.
 * For complex caves or large maps, WASM would be beneficial.
 */
export class CaveGenerator {
  private noise: FastNoiseLite;

  constructor(private config: GeneratorConfig) {
    this.noise = new FastNoiseLite(config.seed + 1000); // Different seed for caves
    this.noise.SetNoiseType(FastNoiseLite.NoiseType.OpenSimplex2);
    this.noise.SetFrequency(0.05);
  }

  /**
   * Generate caves by carving tunnels and caverns
   */
  generate(grid: VoxelGrid): void {
    if (!this.config.generateCaves) {
      console.log('⛰️  Skipping cave generation (disabled)');
      return;
    }

    console.log('🕳️  Generating caves...');
    console.time('caves');

    // Generate worm tunnels
    this.generateWormTunnels(grid);

    // Generate noise-based caverns
    this.generateCaverns(grid);

    console.timeEnd('caves');
    console.log(`  ✓ Caves generated`);
  }

  /**
   * Generate worm-style tunnels that carve through terrain
   */
  private generateWormTunnels(grid: VoxelGrid): void {
    const { caveCount, mapSize, maxHeight } = this.config;
    const random = this.seededRandom(this.config.seed + 2000);

    for (let i = 0; i < caveCount; i++) {
      // Random start position (underground)
      const start: Vector3Int = {
        x: Math.floor(random() * mapSize),
        y: Math.floor(random() * mapSize),
        z: Math.floor(5 + random() * (maxHeight * 0.6))
      };

      this.carveWormTunnel(grid, start, random);
    }
  }

  /**
   * Carve a single worm tunnel
   */
  private carveWormTunnel(grid: VoxelGrid, start: Vector3Int, random: () => number): void {
    const segments = 50 + Math.floor(random() * 50);
    const radius = 2 + Math.floor(random() * 2);

    let pos = { ...start };
    let direction = {
      x: (random() - 0.5) * 2,
      y: (random() - 0.5) * 2,
      z: (random() - 0.5) * 0.5  // Prefer horizontal
    };

    for (let i = 0; i < segments; i++) {
      // Carve sphere at current position
      this.carveSphere(grid, pos, radius);

      // Update direction (smooth random walk)
      direction.x += (random() - 0.5) * 0.5;
      direction.y += (random() - 0.5) * 0.5;
      direction.z += (random() - 0.5) * 0.3;

      // Normalize direction
      const length = Math.sqrt(
        direction.x ** 2 + direction.y ** 2 + direction.z ** 2
      );
      if (length > 0) {
        direction.x /= length;
        direction.y /= length;
        direction.z /= length;
      }

      // Move forward
      pos.x += direction.x * 1.5;
      pos.y += direction.y * 1.5;
      pos.z += direction.z * 1.5;

      // Clamp to bounds
      pos.x = Math.max(1, Math.min(pos.x, grid.width - 2));
      pos.y = Math.max(1, Math.min(pos.y, grid.depth - 2));
      pos.z = Math.max(3, Math.min(pos.z, grid.height - 2));
    }
  }

  /**
   * Carve a sphere (remove voxels)
   */
  private carveSphere(grid: VoxelGrid, center: Vector3Int, radius: number): void {
    const r2 = radius * radius;

    for (let z = -radius; z <= radius; z++) {
      for (let y = -radius; y <= radius; y++) {
        for (let x = -radius; x <= radius; x++) {
          const dist2 = x * x + y * y + z * z;
          if (dist2 <= r2) {
            grid.set({
              x: Math.floor(center.x + x),
              y: Math.floor(center.y + y),
              z: Math.floor(center.z + z)
            }, VoxelType.Air);
          }
        }
      }
    }
  }

  /**
   * Generate noise-based caverns (large open spaces)
   */
  private generateCaverns(grid: VoxelGrid): void {
    const { mapSize, maxHeight, caveThreshold } = this.config;

    // Only check underground voxels (optimization)
    for (let z = 3; z < maxHeight * 0.7; z++) {
      for (let y = 0; y < mapSize; y++) {
        for (let x = 0; x < mapSize; x++) {
          const pos = { x, y, z };

          // Only check solid voxels
          if (!grid.isSolid(pos)) continue;

          // Sample 3D noise
          const noiseValue = this.noise.GetNoise(
            x * 0.1,
            y * 0.1,
            z * 0.1
          );

          // Carve if noise exceeds threshold
          if (noiseValue > caveThreshold) {
            grid.set(pos, VoxelType.Air);
          }
        }
      }
    }
  }

  /**
   * Simple seeded random number generator
   */
  private seededRandom(seed: number): () => number {
    let state = seed;
    return () => {
      state = (state * 1664525 + 1013904223) % 2 ** 32;
      return state / 2 ** 32;
    };
  }
}
